[TOC]


Amplicon-Seq Pipeline
================

The standard MUGQIC Amplicon-Seq pipeline is suited for Metagenome analysis (e.g. ITS, 16S or 18S rDNA analysis).
This pipeline is based on de novo OTU picking method using [Qiime](http://qiime.org). 
First of all, reads are preprocessed by removing trailing low quality with Trimmomatic. Then (assuming we 
have paired end data), the reads are merged with FLASh. Finally a reference chimera dectection is performed in
order to filter out chimeric sequences possibly formed during amplicon sequencing.
Reads are then clustered into different OTU according to a chosen treshold with Usearch. Each OTU is represented
by a single sequence and taxonomic assignation is performed using Uclust. A phylegonetic tree is then built 
with PyNAST which can be useful for beta diversity metrics.
A rarefaction step can be performed in order to compare all the samples and to analyse the alpha and the beta 
diversity. 
Plots for alpha diversity are made with Qiime and Krona tools. Also, Qiime is utilised for beta diversity analysis.

A summary html report is automatically generated by the pipeline. This report contains description
of the sequencing experiment as well as a detailed presentation of the pipeline steps and results.
Various Quality Control (QC) summary statistics are included in the report and additional QC analysis
is accessible for download directly through the report. The report includes also the main references
of the software and methods used during the analysis, together with the full list of parameters
that have been passed to the pipeline main script.

An example of the Amplicon-Seq report for an analysis on public data  is available for illustration purpose only:
[Amplicon-Seq report](http://gqinnovationcenter.com/services/bioinformatics/tools/dnaReport/index.html).

[Here](https://bitbucket.org/mugqic/mugqic_pipelines/downloads/MUGQIC_Bioinfo_DNA-Seq.pptx)
is more information about Amplicon-Seq pipeline that you may find interesting.


Usage
-----
```
#!text

usage: ampliconseq.py [-h] [--help] [-c CONFIG [CONFIG ...]] [-s STEPS]
                 [-o OUTPUT_DIR] [-j {pbs,batch}] [-f] [--clean]
                 [-l {debug,info,warning,error,critical}] [-r READSETS] [-v]

Version: 2.1.0

For more documentation, visit our website: https://bitbucket.org/mugqic/mugqic_pipelines/

optional arguments:
  -h                    show this help message and exit
  --help                show detailed description of pipeline and steps
  -c CONFIG [CONFIG ...], --config CONFIG [CONFIG ...]
                        config INI-style list of files; config parameters are
                        overwritten based on files order
  -s STEPS, --steps STEPS
                        step range e.g. '1-5', '3,6,7', '2,4-8'
  -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                        output directory (default: current)
  -j {pbs,batch}, --job-scheduler {pbs,batch}
                        job scheduler type (default: pbs)
  -f, --force           force creation of jobs even if up to date (default:
                        false)
  --clean               create 'rm' commands for all job removable files in
                        the given step range, if they exists; if --clean is
                        set, --job-scheduler, --force options and job up-to-
                        date status are ignored (default: false)
  -l {debug,info,warning,error,critical}, --log {debug,info,warning,error,critical}
                        log level (default: info)
  -r READSETS, --readsets READSETS
                        readset file
  -v, --version         show the version information and exit

Steps:
------
1- trimmomatic
2- merge_trimmomatic_stats
3- flash
4- merge_flash_stats
5- catenate
6- catenate_split
7- uchime
8- merge_uchime
9- filter_chimeras
10- otu_picking
11- otu_rep_picking
12- otu_assigning
13- otu_table
14- merge_uchime_stats
15- otu_alignment
16- filter_alignment
17- phylogeny
18- qiime_report
19- multiple_rarefaction
20- alpha_diversity
21- collate_alpha
22- sample_rarefaction_plot
23- qiime_report2
24- single_rarefaction
25- rarefaction_plot
26- summarize_taxa
27- plot_taxa
28- plot_heatmap
29- krona
30- plot_to_alpha
31- beta_diversity
32- pcoa
33- pcoa_plot
34- plot_to_beta

```
1- trimmomatic
--------------
Raw reads quality trimming and removing of Illumina adapters is performed using [Trimmomatic](http://www.usadellab.org/cms/index.php?page=trimmomatic).
If an adapter FASTA file is specified in the config file (section 'trimmomatic', param 'adapter_fasta'),
it is used first. Else, 'Adapter1' and 'Adapter2' columns from the readset file are used to create
an adapter FASTA file, given then to Trimmomatic. For PAIRED_END readsets, readset adapters are
reversed-complemented and swapped, to match Trimmomatic Palindrome strategy. For SINGLE_END readsets,
only Adapter1 is used and left unchanged.

This step takes as input files:

1. FASTQ files from the readset file if available
2. Else, FASTQ output files from previous picard_sam_to_fastq conversion of BAM files

2- merge_trimmomatic_stats
--------------------------
The trim statistics per readset are merged at this step.

3- flash
--------
Merge paired end reads using [FLASh](http://ccb.jhu.edu/software/FLASH/).

4- merge_flash_stats
--------------------
The paired end merge statistics per readset are merged at this step.

5- catenate
-----------
Catenate all the reads in one file for further analysis.

This step takes as input files:

1. Merged FASTQ files from previous step flash. 

6- catenate_split
-----------------
Split the catenate file for UCHIME.

7- uchime
---------
Reference based chimera detection is performed using [Usearch/Uchime](http://drive5.com/usearch/) (http://drive5.com/usearch/manual/uchime_algo.html).

This step takes as input files:

1. Catenated FASTA file from previous step catenate. 

8- merge_uchime
---------------
Merge the chimeras files from previous step.

9- filter_chimeras
------------------
Filter the input catenate data file by passing the chimeras file created in the previous step.

This step takes as input files:

1. Catenated FASTQ file from 3rd step cat_data.
2. Chimera file from previous step uchime.

10- otu_picking
---------------
The OTU picking step (de novo) assigns similar sequences to operational taxonomic units (OTUs) by clustering sequences based on a user-defined similarity threshold. Method per default uses [sumaclust] (http://www.grenoble.prabi.fr/trac/sumatra/wiki/documentation/sumaclust) program wrapped by [Qiime] (http://qiime.org).

This step takes as input file:

1. Catenated and filtered FASTA file from previous step.


11- otu_rep_picking
-------------------
After picking OTUs, this step pick a representative sequence for each OTU.

This step takes as input files:

1. OTU file from previous step 
2. Catenated and filtered FASTA file from filter_chimeras step.


12- otu_assigning
-----------------
Given a set of OTUS, this step attempts to assign the taxonomy of each OTU using [Uclust] (http://drive5.com/usearch/manual/uclust_algo.html).

This step takes as input files:

1. OTU representative sequence file from previous step.


13- otu_table
-------------
This step make a consensus OTU table in biom format. It tabulates the number of times an OTU is found in each sample, and adds the taxonomic predictions for each OTU. 

This step takes as input files:

1. OTU picking file.
2. Taxonomy assignment for each OTU from the previous step.


14- merge_uchime_stats
----------------------
The chimeric sequences filtered out statistics per readset are merged at this step.

15- otu_alignment
-----------------
Align OTU representative sequences using [PyNAST] (http://biocore.github.io/pynast/).

This step takes as input file:

1. OTU representative sequence file.


16- filter_alignment
--------------------
Filter the alignment by removing positions which are gaps in every sequence.

This step takes as input file:

1. Alignment sequence file.


17- phylogeny
-------------
Build a phylogenetic tree from a multiple sequence alignment using [FastTree] (http://www.microbesonline.org/fasttree/).

This step takes as input file:

1. Filtered alignment sequence file from previous step.


18- qiime_report
----------------
1st part report for taxonomic affiliation. 

19- multiple_rarefaction
------------------------
1st step (/4) for rarefaction plot.
Rarefies OTU table by random sampling (without replacement) at different depth in order to perform rarefaction analysis. 
You need to provide the minimum/maximum number of sequences per samples and the size of each steps between the min/max of seqs/sample. 

This step takes as input files:

1. OTU rarefied table in biom format if available.
2. Else, OTU non rarefied table in biom format.


20- alpha_diversity
-------------------
2nd step (/4) for rarefaction plot.
Calculate alpha diversity on each sample using a variety of alpha diversity metrics (chao1, shannon, observed otus). 

This step takes as input files:

1. Multiple OTU rarefied table in biom format from previous step.


21- collate_alpha
-----------------
3rd step (/4) for rarefaction plot.
Merge all the alpha diversity computed in the previous step. 

22- sample_rarefaction_plot
---------------------------
Last step for rarefaction plot.
Plot the rarefaction curve for each sample

23- qiime_report2
-----------------
2nd part report for taxonomic affiliation. Plot rarefaction curve for each sample.

24- single_rarefaction
----------------------
This step is recommended. It subsamples (rarefy) all the samples to an equal number of sequences for further comparaison.
You have to provide the number of sequences to subsample per sample in the configuration file (single_rarefaction_depth).

This step takes as input files:

1. OTU table in biom format.


25- rarefaction_plot
--------------------
Last step for rarefaction plot.
Plot the rarefaction curve for rarefied data. 

26- summarize_taxa
------------------
1st step (/3) for taxonomic affiliation plot.
Summarize information of taxonomic groups within each sample at different taxonomic level. 

This step takes as input files:

1. OTU rarefied table in biom format if available.
2. Else, OTU non rarefied table in biom format.


27- plot_taxa
-------------
2nd step (/3) for taxonomic affiliation plot.
Make taxaonomy summary bar plots based on taxonomy assignment. 

This step takes as input files:

1. Summarized information from previous step.


28- plot_heatmap
----------------
Last step for taxonomic affiliation plot.
Make heatmap at phylum level. 

This step takes as input files:

1. Summarized information from previous step.


29- krona
---------
Plot Krona chart for taxonomic affiliation

30- plot_to_alpha
-----------------
Final report 1st part for the Amplicon-Seq pipeline. Display results (taxonomy, heatmap and alpha diversity).

31- beta_diversity
------------------
1st step (/3) for 2D PCoA plot.
Calculate beta diversity (pairwise sample dissimilarity) on OTU table. The OTU table has to be rarefied. 
Only works with >= 4 samples

This step takes as input files:

1. OTU rarefied table in biom format.
2. Tree file.


32- pcoa
--------
2nd step (/3) for 2D PCoA plot.
Compute coordinates pour PCoA 

This step takes as input file:

1. Matrix produced in the previous step.


33- pcoa_plot
-------------
Last step for 2D PCoA plot.

This step takes as input file:

1. PCoA from the previous step.


34- plot_to_beta
----------------
Final report's 2nd part for the Amplicon-Seq pipeline. Display results (beta diversity PCoA plots).
