```{r preparation, results="hide",echo=F,message=F,warning=F}

### DEFINED OUTSIDE
# report_dir

library(rmarkdown)
library(knitr)
library(hwriter)
library(magrittr)
library(jsonlite)
options(stringsAsFactors=F)


# Directories and cp data
unlink( file.path(report_dir,"kallisto") , recursive = T)
dir.create( file.path(report_dir,"kallisto"), showWarnings=F,recursive=T)
file.copy(from = "kallisto", to = report_dir, overwrite = T, recursive = T)

# Create summary table
all_readsets=read.delim(file.path(report_dir,"kallisto/All_readsets/all_readsets.abundance_genes.csv"))
readset_names=colnames(all_readsets)[-1]

mat=data.frame(matrix(nrow=length(readset_names), ncol=8))
rownames(mat)=readset_names
colnames(mat)=c("Readset","Raw Paired Reads #", "Transcripts #", "Sum transcripts","%", "Genes #", "Sum genes","%")
mat[,1]=readset_names
for (i in seq(length(readset_names))){
    readset=readset_names[i]
    run_info=fromJSON(file.path(report_dir, "kallisto", readset, "run_info.json"))
    mat[i,"Raw Paired Reads #"]=run_info$n_processed
    mat[i,"Transcripts #"]=run_info$n_targets
    transcript_count=read.delim(file.path(report_dir,"kallisto",readset,"abundance_transcripts.tsv"))
    mat[i,"Sum transcripts"]=sum(transcript_count[,"est_counts"])
    mat[i,5]=signif(mat[i,"Sum transcripts"]/mat[i,"Raw Paired Reads #"],3)*100
    gene_count=read.delim(file.path(report_dir,"kallisto",readset,"abundance_genes.tsv"))
    mat[i,"Genes #"]=nrow(gene_count)
    mat[i,"Sum genes"]=sum(gene_count$counts)
    mat[i,8]=signif(mat[i,"Sum genes"]/mat[i,"Raw Paired Reads #"],3)*100
}
#mat=apply(mat, 1, function(x){prettyNum(x,big.mark=",",scientific=FALSE)})

```

### Kallisto

[Kallisto](https://pachterlab.github.io/kallisto/about) is a program for quantifying abundances of transcripts from RNA-Seq data, or more generally of target sequences using high-throughput sequencing reads. It is based on the novel idea of pseudoalignment for rapidly determining the compatibility of reads with targets, without the need for alignment.

kallisto is ran fastq, producing raw count
transcripts are then merged to give the count at the gene level

transcripts2genes file: ([file](Mus_musculus.GRCm38.Ensembl83.tx2gene))

Gene level count were obtained using the tximport package v.ZZZ.

```{r summary_table, results="markup",echo=F,message=F,warning=F}
kable(mat,align="r",row.names=F, format.args = list(decimal.mark = ".", big.mark = ","))

```


"Readset","Raw Reads #", "Trancriptome", "Transcripts", "Reads transcripts #","% Read used", "Genes", "Reads genes #","% Read used"




* Raw Reads: total number of reads obtained from the sequencer
* Trancriptome: number of transcript targets
* Transcripts: number of Transcripts with at least 5 reads
* Reads transcripts #: number of aligned reads to a transcripts
* % Read used: Reads transcripts # / Surviving reads
* Alternative Alignments: number of duplicate read entries providing alternative coordinates
* %: Alternative Alignments / Aligned Reads
* rRNA Reads: number of reads aligning to rRNA regions as defined in the transcript model definition
* %: rRNA Reads / Surviving Reads
* Coverage: mean coverage = number of bp aligning to reference / total number of bp in the reference
* Exonic Rate: fraction mapping reads within exons

* Genes: number of Genes with at least 5 reads
