RNA-Seq Pipeline
================


Overview
--------
The standard MUGQIC RNA-Seq pipeline uses [TopHat](http://tophat.cbcb.umd.edu/) to align reads to the reference genome and discover transcript splice sites. These alignments are used during downstream analysis in several ways. [Cufflinks](http://cufflinks.cbcb.umd.edu/) uses this map against the reference genome to assemble the reads into transcripts. Cufflinks is executed in two different ways:

* Cufflinks_denovo: Cufflinks construct a parsimonious set of transcripts that "explain" the reads observed in the RNA-Seq experiment.
* Cufflinks_known: A reference annotation file is supplied to estimate isoform expression. Cufflinks will not assemble novel transcripts, and the program will ignore alignments not structurally compatible with any reference transcript.

Cuffdiff, a part of the Cufflinks package, takes the aligned reads from two conditions and reports transcripts that are differentially expressed. FPKM values calculated by cufflinks are used as input of the differential transcript expression analysis. This differences are reported in the isoform_exp.diff.with.fpkm.csv files located in the cuffdiff_denovo and cuffdiff_known directories, one directory per design. Otherwise, differential gene expression uses raw read counts generated by HTSeq as input. The *DESeq* and *edgeR* methods model **count data** by a negative binomial distribution. The parameters of the distribution (mean and dispersion) are estimated from the data, i.e. from the read counts in the input files. Both methods compute a measure of read abundance, i.e. expression level (called *base mean* or *mean of normalized counts* in DESeq, and *concentration* in edgeR) for each gene and apply a hypothesis test to each gene to evaluate differential expression. In particular, both methods determine a p-value and a log2 fold change (in expression level) for each gene. The Log2 FC of EdgeR is reported in the differential gene results file, one file per design.

The log2fold change is the logarithm (to basis 2) of the fold change condition from condition A to B (mutation or treatment are the most common conditions). A “fold change” between conditions A and B at a gene or transcript is normally computed as the ratio at gene or transcript of the base mean of scaled counts for condition B to the base mean of scaled counts for condition A. Counts are scaled by a size factor in a step called normalisation (if the counts of non-differentially expressed genes in one sample are, on average, twice as high as in another,  the size factor for the first sample should be twice that of the other sample).  Each column of the count table is then divided by the size factor for this column and the count values are brought to a common scale, making them comparable. See the [EdgeR vignette](http://www.bioconductor.org/packages/2.12/bioc/vignettes/edgeR/inst/doc/edgeR.pdf) for additional information on normalization approaches used in the pipeline.

 

A summary html report is automatically generated by the pipeline. This report contains description of the sequencing experiment as well as a detailed presentation of the pipeline steps and results. Various Quality Control (QC) summary statistics are included in the report and additional QC analysis is accessible for download directly through the report. The report includes also the main references of the software and methods used during the analysis, together with the full list of parameters that have been passed to the pipeline main script.

An example of the RNAseq report for an analysis on Public Corriel CEPH B-cell is available for illustration purpose only: [RNaseq report](http://gqinnovationcenter.com/services/bioinformatics/tools/rnaReport/index.html).

[Here](https://bitbucket.org/mugqic/mugqic_pipeline/downloads/MUGQIC_Bioinfo_RNA-Seq.pptx) is more information about RNASeq pipeline that you may find interesting.


On this page:

[TOC]

Usage
-----
```
#!bash
mugqic_pipeline/pipelines/rnaseq/rnaseq.py --help
```


Steps
-----
The standard differential expression analysis for RNAseq data performs the following steps:

### trimming
Raw reads quality trimming and removing of Illumina adapters is performed using trimmomatic. 

### trimMetrics
Generates the trimming statistics file

### aligning
The filtered reads are aligned to a reference genome. The alignment is done per lane of sequencing using the combination of tophat/bowtie software. It generates a Binary Alignment Map file (.bam).

### merging
Bam files per sample are merged in one file. Merge is done using the Picard software. The resulting alignment file is reordered (karyotypic) and resulting reads per sample are marked as duplicates if they have the same 5' alignment positions (for both mates in the case of paired-end reads). All but the best pair (based on alignment score) will be marked as a duplicate in the .bam file. Marking duplicates and reorder are executed using the Picard software.

### wiggle
Generate wiggle tracks suitable for multiple browsers.

### rawCounts
Counting reads in features using htseq-count.

### rawCountsMetrics
Create rawcount matrix, zip the wiggle tracks and create the saturation plots based on standardized read counts.

### fpkm
Compute FPKM measures for de novo and known transcripts using cufflinks.

### exploratory
ExploratoryAnalysis using the gqSeqUtils R package

### cuffdiff
The transcript quantification engine of Cufflinks (Cuffdiff) is used to calculate transcript expression levels in more than one condition and test them for signficant differences.

### dge
Differential gene expression analysis using DESEQ and EDGER. Merge the results of the analysis in a single csv file.

### goseq
Gene Ontology analysis for RNA-seq using the bioconductor's R package goseq. Generates Go annotations for cuffdiff known transcripts differential expression and differential gene expression analysis.

### deliverable
Generates the standard report. A summary html report contains the description of the sequencing experiment as well as a detailed presentation of the pipeline steps and results. Various Quality Control (QC) summary statistics are included in the report and additional QC analysis is accessible for download directly through the report. The report includes also the main references of the software and methods used during the analysis, together with the full list of parameters passed to the pipeline main script.
