DNA-Seq Pipeline
================


Overview
--------
The standard MUGQIC DNA-Seq pipeline uses BWA to align reads to the reference genome. Treatment and filtering of mapped reads approaches as INDEL realignment, mark duplicate reads, recalibration and sort are executed using Picard and GATK. Samtools MPILEUP and bcftools are used to produce the standard SNP and indels variants file (VCF). Additional SVN annotations mostly applicable to human samples include mappability flags, dbSNP annotation and extra information about SVN by using published databases.  The SNPeff tool is used to annotate variants using an integrated database of functional predictions from multiple algorithms (SIFT, Polyphen2, LRT and MutationTaster, PhyloP and GERP++, etc.) and to calculate the effects they produce on known genes. A list of effects and annotations that SnpEff calculate can be found [here](http://snpeff.sourceforge.net/faq.html#What_effects_are_predicted?).

A summary html report is automatically generated by the pipeline. This report contains description of the sequencing experiment as well as a detailed presentation of the pipeline steps and results. Various Quality Control (QC) summary statistics are included in the report and additional QC analysis is accessible for download directly through the report. The report includes also the main references of the software and methods used during the analysis, together with the full list of parameters that have been passed to the pipeline main script.

An example of the DNA-Seq report for an analysis on public data  is available for illustration purpose only: [DNA-Seq report](http://gqinnovationcenter.com/services/bioinformatics/tools/dnaReport/index.html).

[Here](https://bitbucket.org/mugqic/mugqic_pipelines/downloads/MUGQIC_Bioinfo_DNA-Seq.pptx) is more information about DNA-Seq pipeline that you may find interesting.


On this page:

[TOC]

Usage
-----
```
#!bash
mugqic_pipelines/pipelines/dnaseq/dnaseq.py --help
```


Steps
-----

The standard variant discovery pipeline performs the following steps:

### trimAndAlign
Raw reads quality trimming and removing of Illumina adapters is performed using trimmomatic. The filtered reads are aligned to a reference genome. The alignment is done per lane of sequencing. The alignment software used is bwa, two different algorithms are available: bwa aln (backtrack) and bwa mem.

### mergeTrimStats
The trim statistics per lane/sample are merged at this step.

### mergeLanes
Bam files per sample are merged in one file. Merge is done using the Picard software.

### indelRealigner
Insertion and deletion realignment is performed on regions where multiple base mismatches are preferred over indels by the aligner since it can appear to be less costly by the algorithm. Such regions will introduce false positive variant calls which may be filtered out by realigning those regions properly. Realignment is done using the GATK software. The reference genome is divided by a number regions given by the nbRealignJobs parameter.

### mergeRealigned
Merging the regions of realigned reads per sample is done using Picard.

### fixmate
Fixing the read mates. Once local regions are realigned, the read mate coordinates of the aligned reads need to be recalculated since the reads are realigned at positions that differ from their original alignment. Fixing the read mate positions is done using the Picard software.

### markDup
Marking duplicates. Aligned reads per sample are duplicates if they have the same 5' alignment positions (for both mates in the case of paired-end reads). All but the best pair (based on alignment score) will be marked as a duplicate in the .bam file. Marking duplicates is done using the Picard software.

### recalibration
Recalibrate base quality scores of sequencing-by-synthesis reads in an aligned BAM file. After recalibration, the quality scores in the QUAL field in each read in the output BAM are more accurate in that the reported quality score is closer to its actual probability of mismatching the reference genome. Moreover, the recalibration tool attempts to correct for variation in quality with machine cycle and sequence context, and by doing so provides not only more accurate quality scores but also more widely dispersed ones.

### metrics
Compute metrics and generate coverage tracks per sample. Multiple metrics are computed at this stage:Number of raw reads, Number of filtered reads, Number of aligned reads, Number of duplicate reads, Median, mean and standard deviation of insert sizes of reads after alignment, mean coverage over exons (mean number of reads per base position), percentage of bases covered at X reads (%_bases_above_50 means the % of exons bases which have at least 50 reads) whole genome
percentage of bases covered at X reads (%_bases_above_50 means the % of exons bases which have at least 50 reads) for specific targets (CCDS regions are used in human samples). A TDF (.tdf) coverage track is also generated at this step for easy visualization of coverage in the IGV browser.

### calculateHSMetrics
Compute on target percent of hybridasation based capture

### callableBases
Computes the callable region or the genome as a bed track

### extractCommonSNPFreq
Extracts allele frequecies of possible variants accross the genome

### BAFPlot
Plots DepthRatio and B allele frequency of previously extracted alleles

### haplotypeCaller
GATKs haplotype caller for snps and small indels

### metricsLibrarySample
Merges the gvcfs of haplotype caller and also generates a per sample vcf containing genotypes.

### metricsLibrarySample
Merge metrics. Read metrics per sample are merged at this step.

### snpAndIndelBCF
Mpileup and Variant calling. Variants (SNPs and INDELs) are called using samtools mpileup. bcftools view is used to produce binary bcf files.  

### mergeFilterBCF
bcftools is used to merge the raw binary variants files created in the snpAndIndelBCF step.  The output of bcftools is fed to varfilter, which does an additional filtering of the variants and transforms the output into the VCF (.vcf) format. One vcf file contain the SNP/INDEL calls for all samples in the experiment. 

### filterNStretches
The final .vcf files are filtered for long 'N' INDELs which are sometimes introduced and cause excessive memory usage by downstream tools.

### flagMappability
Mappability annotation. An in-house database identifies regions in which reads are confidently mapped to the reference genome.

### snpIDAnnotation
dbSNP annotation. The .vcf files are annotated for dbSNP using the software SnpSift (from the SNPEff suite).

### snpEffect
Variant effect annotation. The .vcf files are annotated for variant effects using the SnpEff software. SnpEff annotates and predicts the effects of variants on genes (such as amino acid changes).

### dbNSFPAnnotation
Additional SVN annotations. Provides extra information about SVN by using numerous published databases. Applicable to human samples. Databases available include Biomart (adds GO annotations based on gene information) and dbNSFP (an integrated database of functional annotations from multiple sources for the comprehensive collection of human non-synonymous SNPs. It compiles prediction scores from four prediction algorithms (SIFT, Polyphen2, LRT and MutationTaster), three conservation scores (PhyloP, GERP++ and SiPhy) and other function annotations) .

### metricsSNV
Metrics SNV. Multiple metrics associated to annotations and effect prediction are generated at this step: change rate by chromosome, changes by type, effects by impact, effects by functional class, counts by effect, counts by genomic region, SNV quality, coverage, InDel lengths, base changes,  transition-transversion rates, summary of allele frequencies, codon changes, amino acid changes, changes per chromosome, change rates.

### deliverable
Generating the standard report. A summary html report is automatically generated by the pipeline. This report contains description of the sequencing experiment as well as a detailed presentation of the pipeline steps and results. Various Quality Control (QC) summary statistics are included in the report and additional QC analysis is accessible for download directly through the report. The report includes also the main references of the software and methods used during the analysis, together with the full list of parameters passed to the pipeline main script.

### fullPileup
Full pileup (optional). A raw mpileup file is created using samtools mpileup and compressed in gz format. One packaged mpileup file is created per sample/chromosome.
